File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* helper(vector<int>& preorder, vector<int>& inorder, int bp, int ep, int bi, int ei){
        if(bp>ep) return NULL;
        int rVal=preorder[bp];
        int idx=bi;
        while(inorder[idx]!=rVal) ++idx;
        int lft=idx-bi;
        TreeNode *root= new TreeNode(rVal);
        root->left=helper(preorder, inorder, bp+1, bp+lft, bi, bi+lft-1);
        root->right=helper(preorder, inorder, bp+lft+1, ep, bi+lft+1, ei);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if(preorder.size()==0) return NULL;
        return helper(preorder, inorder, 0, (int)preorder.size()-1, 0, (int)inorder.size()-1);
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int preIdx;
    Map<Integer,Integer> map;
    private TreeNode buildTreeUtil(int[] preorder, int lo, int hi){
        if(lo>hi) return null;
        TreeNode root= new TreeNode(preorder[preIdx++]);
        int ind= map.get(root.val);
        root.left= buildTreeUtil(preorder, lo, ind-1);
        root.right= buildTreeUtil(preorder, ind+1, hi);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n= inorder.length;
        map= new HashMap<>();
        for(int i=0;i<n;i++) map.put(inorder[i],i);
        preIdx=0;
        return buildTreeUtil(preorder, 0, n-1);
    }
}